File: java-devildyw/src/main/java/top/dyw/leetcode/TOP面试题/L13罗马数字转整数.java
package top.dyw.leetcode.TOP面试题;

import java.util.HashMap;

/**
 * @author Devil
 * @since 2022-12-08-14:03
 */
@SuppressWarnings("all")
public class L13罗马数字转整数 {

    public static void main(String[] args) {
        System.out.println(new L13罗马数字转整数().romanToInt("MCMXCIV"));
    }
    public int romanToInt(String s) {
        HashMap<String,Integer> map = new HashMap<>();
        map.put("I",1);
        map.put("IV",4);
        map.put("V",5);
        map.put("IX",9);
        map.put("X",10);
        map.put("XL",40);
        map.put("L",50);
        map.put("XC",90);
        map.put("C",100);
        map.put("CD",400);
        map.put("D",500);
        map.put("CM",900);
        map.put("M",1000);

        int sum = 0;
        for (int i = 0; i < s.length(); i++) {
            if (i+1<s.length()&&map.containsKey(s.substring(i,i+2))){
                sum+=map.get(s.substring(i,i+2));
                i++;
            }else{
                sum+=map.get(s.substring(i,i+1));
            }
        }
        return sum;
    }
}

File: golang-devildyw/simple/L13罗马数字转整数.go
package main

func main() {
	print(romanToInt("MCMXCIV"))
}

func romanToInt(s string) int {
	var (
		n                       = len(s)
		romanMap map[string]int = make(map[string]int)
		result   int            = 0
	)

	romanMap["I"] = 1
	romanMap["V"] = 5
	romanMap["X"] = 10
	romanMap["L"] = 50
	romanMap["C"] = 100
	romanMap["D"] = 500
	romanMap["M"] = 1000
	romanMap["IV"] = 4
	romanMap["IX"] = 9
	romanMap["XL"] = 40
	romanMap["XC"] = 90
	romanMap["CD"] = 400
	romanMap["CM"] = 900

	for i := 0; i < n; i++ {
		if i+1 < n {
			if value, ok := romanMap[s[i:i+2]]; ok {
				result += value
				i++
				continue
			}
		}
		result += romanMap[s[i:i+1]]

	}
	return result
}
